components/common/src/main/java/com/hotels/styx/common/FailureHandlingTaskProcessor.java

```diff
@@ -0,0 +1,144 @@
+     * @return a list of outputsl
+     */
+    public List<R> process(List<T> inputs, Task<T, R> task) {
+        List<R> successes = new ArrayList<>();
+        Map<T, Exception> failures = new LinkedHashMap<>();
+
+        inputs.forEach(input -> {
+            try {
+                successes.add(task.execute(input));
+            } catch (Exception t) {
+                onEachFailure.accept(input, t);
+                failures.put(input, t);
+            }
+        });
+
+        if (!failures.isEmpty()) {
+            failuresPostProcessing.accept(failures);
+        }
+
+        return successes;
+    }
+
+    /**
+     * Builds {@link FailureHandlingTaskProcessor}.
+     *
+     * @param <T> input type
+     * @param <R> output type
+     */
+    public static final class Builder<T, R> {
+        // Logs by default so that failures do not become invisible.
+        private BiConsumer<T, Exception> onEachFailure = (input, err) -> LOGGER.error("Failed on input " + input, err);
+        private Consumer<Map<T, Exception>> failuresPostProcessing = failures -> LOGGER.error("Failures: " + failures);
+
+        /**
+         * An action to take immediately after a failure. This can be used to fail fast by rethrowing the error,
+         * or to log the failure as soon as it happens without waiting.
+         *
+         * @param onEachFailure an action to take
+         * @return this builder
+         */
+        public Builder<T, R> doImmediatelyOnEachFailure(BiConsumer<T, Exception> onEachFailure) {
+            this.onEachFailure = requireNonNull(onEachFailure);
+            return this;
+        }
+
+        /**
+         * An action to take after processing all list items, if there were failures.
+         * This will only execute if the individual failures were did not rethrow their exceptions.
+         *
+         * This action may be used to throw an exception, otherwise the {@link #process(List, Task)} method
+         * will return whatever successful results it had (or an empty list if none).
+         *
+         * @param failuresPostProcessing an action to take
+         * @return this builder
+         */
+        public Builder<T, R> doOnFailuresAfterAllProcessing(Consumer<Map<T, Exception>> failuresPostProcessing) {
+            this.failuresPostProcessing = requireNonNull(failuresPostProcessing);
+            return this;
+        }
+
+        /**
+         * Build an instance of {@link FailureHandlingTaskProcessor} using the configuration provided.
+         *
+         * @return a new instance
+         */
+        public FailureHandlingTaskProcessor<T, R> build() {
+            return new FailureHandlingTaskProcessor<>(this);
+        }
+    }
+
+    /**
+     * A task to execute. Similar to {@link java.util.function.Function}, except it is allowed to throw any type of
+     * {@link Exception}.
+     *
+     * @param <T> input type
+     * @param <R> output type
+     */
+    public interface Task<T, R> {
```

Comment:
If I were to read this the code for the first time, I would probably interpret 'Task' as something to do with concurrency... If you agree it could be misleading, can you think of a better name?